File: Tools/tax_calculator.py
from decimal import Decimal, ROUND_HALF_UP, InvalidOperation # decimal module for precise decimal arithmetic typically use for financial calculations

def safe_decimal(value, default=0):
    """
    Safely convert a value to a Decimal.
    
    Args:
        value: The value to convert to a Decimal.
        default: The default value to use if conversion fails.
    
    Returns:
        A Decimal representation of the value or the default value.
    """
    try:
        return Decimal(str(value or default))
    except (ValueError, TypeError, InvalidOperation):
        return Decimal(str(default))

def calculate_tax(price, tax_rate):
    """Calculate tax amount with precise decimal arithmetic"""
    price_dec = Decimal(str(price))
    rate_dec = Decimal(str(tax_rate)) / Decimal('100')
    return price_dec * rate_dec

def validate_vat_rate(vat_rate):
    """
    Validate VAT rate input.

    Args:
        vat_rate: The VAT rate to validate (can be string, int, float, or Decimal)

    Returns:
        Decimal: Validated VAT rate

    Raises:
        ValueError: If VAT rate is invalid (negative, non-numeric, or out of reasonable range)
    """
    try:
        # Try to convert directly to Decimal to catch invalid inputs
        rate = Decimal(str(vat_rate)) if vat_rate not in (None, '') else Decimal('0')

        if rate < Decimal('0'):
            raise ValueError("VAT rate cannot be negative")

        if rate > Decimal('100'):
            raise ValueError("VAT rate cannot exceed 100%")

        return rate
    except (InvalidOperation, ValueError) as e:
        if "VAT rate cannot" in str(e):
            raise
        raise ValueError(f"Invalid VAT rate: must be a numeric value between 0 and 100")
    except TypeError:
        raise ValueError(f"Invalid VAT rate: must be a numeric value between 0 and 100")

def calculate_vat(data):
    """
    Calculate VAT (Value Added Tax) for international markets.

    This function handles VAT-exclusive pricing where tax is calculated on the net price
    and added to get the gross (total) price.

    Args:
        data: A dictionary containing the following keys:
            - items: List of item dictionaries with 'price'
            - vat_rate: The VAT percentage to apply (e.g., 12.00 for Philippines, 20.00 for EU)
            - discounts: (Optional) List of discount dictionaries with 'amount'
            - shipping_cost: (Optional) The cost of shipping
            - shipping_taxable: (Optional) Boolean indicating if shipping is subject to VAT
            - is_sales_before_tax: (Optional) Boolean indicating if discounts are applied before tax calculation
            - discount_is_taxable: (Optional) Boolean indicating if discounts themselves are taxable

    Returns:
        A dictionary containing:
            - net_amount: Total before VAT (items + shipping - discounts)
            - vat_amount: Total VAT charged
            - gross_amount: Total including VAT (net + vat)
            - vat_rate_applied: The VAT rate that was applied
            - item_total: Total amount of items before VAT
            - discount_total: Total amount of discounts
            - shipping_cost: Shipping cost
            - vat_breakdown: List of VAT breakdown items

    Raises:
        ValueError: If vat_rate is invalid

    Example:
        data = {
            'items': [{'price': 100}, {'price': 50}],
            'vat_rate': 12,
            'discounts': [{'amount': 10}],
            'shipping_cost': 5,
            'shipping_taxable': True,
            'is_sales_before_tax': False,
            'discount_is_taxable': True
        }
        result = calculate_vat(data)
        # Returns: {
        #     'net_amount': 145.00,
        #     'vat_amount': 17.40,
        #     'gross_amount': 162.40,
        #     'vat_rate_applied': 12.00,
        #     ...
        # }
    """
    # Validate VAT rate
    vat_rate = validate_vat_rate(data.get('vat_rate', 0))

    # Extract data
    items = data.get('items', [])
    discounts = data.get('discounts', [])
    shipping_cost = safe_decimal(data.get('shipping_cost', 0))
    shipping_taxable = data.get('shipping_taxable', True)  # VAT typically applies to shipping
    is_sales_before_tax = data.get('is_sales_before_tax', False)
    discount_is_taxable = data.get('discount_is_taxable', True)

    # Initialize totals
    item_total = Decimal('0')
    discount_total = Decimal('0')
    net_amount = Decimal('0')
    vat_amount = Decimal('0')
    vat_breakdown = []

    # Calculate item totals
    for item in items:
        price = safe_decimal(item.get('price', 0))
        item_total += price

    # Calculate discount totals
    for discount in discounts:
        discount_total += safe_decimal(discount.get('amount', 0))

    # Calculate VAT based on discount behavior settings
    if is_sales_before_tax:
        # Apply discounts before calculating VAT
        taxable_items_amount = item_total - discount_total
        
        if discount_is_taxable:
            # Condition 2: Discounts applied before tax, but discounts are taxable
            # Tax on (items - discounts) + tax on discounts
            if taxable_items_amount > 0 and vat_rate > 0:
                items_vat = calculate_tax(taxable_items_amount, vat_rate)
                vat_amount += items_vat
                vat_breakdown.append({
                    'item': 'Items',
                    'net_amount': float(taxable_items_amount.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
                    'vat': float(items_vat.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
                })
            
            if discount_total > 0 and vat_rate > 0:
                discount_vat = calculate_tax(discount_total, vat_rate)
                vat_amount += discount_vat
                vat_breakdown.append({
                    'item': 'Discounts',
                    'net_amount': float(discount_total.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
                    'vat': float(discount_vat.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
                })
        else:
            # Condition 4: Discounts applied before tax, discounts not taxable
            # Tax only on (items - discounts)
            if taxable_items_amount > 0 and vat_rate > 0:
                items_vat = calculate_tax(taxable_items_amount, vat_rate)
                vat_amount += items_vat
                vat_breakdown.append({
                    'item': 'Items',
                    'net_amount': float(taxable_items_amount.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
                    'vat': float(items_vat.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
                })
    else:
        # Apply discounts after calculating VAT
        if discount_is_taxable:
            # Condition 1: Discounts applied after tax, discounts are taxable
            # Tax on items + tax on discounts
            if item_total > 0 and vat_rate > 0:
                items_vat = calculate_tax(item_total, vat_rate)
                vat_amount += items_vat
                vat_breakdown.append({
                    'item': 'Items',
                    'net_amount': float(item_total.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
                    'vat': float(items_vat.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
                })
            
            if discount_total > 0 and vat_rate > 0:
                discount_vat = calculate_tax(discount_total, vat_rate)
                vat_amount += discount_vat
                vat_breakdown.append({
                    'item': 'Discounts',
                    'net_amount': float(discount_total.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
                    'vat': float(discount_vat.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
                })
        else:
            # Condition 3: Discounts applied after tax, discounts not taxable
            # Tax only on items
            if item_total > 0 and vat_rate > 0:
                items_vat = calculate_tax(item_total, vat_rate)
                vat_amount += items_vat
                vat_breakdown.append({
                    'item': 'Items',
                    'net_amount': float(item_total.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
                    'vat': float(items_vat.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
                })

    # Calculate net amount for display
    if is_sales_before_tax and not discount_is_taxable:
        # For condition 4, net amount is already reduced by discounts
        net_amount = item_total - discount_total
    else:
        # For other conditions, net amount shows items before discounts
        net_amount = item_total - discount_total

    # Add shipping to net amount
    net_amount += shipping_cost

    # Calculate VAT on shipping if applicable
    if shipping_cost > 0 and shipping_taxable and vat_rate > 0:
        shipping_vat = calculate_tax(shipping_cost, vat_rate)
        vat_amount += shipping_vat
        vat_breakdown.append({
            'item': 'Shipping',
            'net_amount': float(shipping_cost.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
            'vat': float(shipping_vat.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))
        })

    # Calculate gross amount (total including VAT)
    gross_amount = net_amount + vat_amount

    # Return results with proper rounding
    result = {
        'item_total': float(item_total.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'discount_total': float(discount_total.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'shipping_cost': float(shipping_cost.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'net_amount': float(net_amount.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'vat_amount': float(vat_amount.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'gross_amount': float(gross_amount.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'vat_rate_applied': float(vat_rate.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'vat_breakdown': vat_breakdown
    }

    return result

def tax_calculator(data):
    """
    Calculate tax based on provided data.
    
    Args:
        data: A dictionary containing the following keys:
            - items: List of item dictionaries with 'price' and 'tax_rate'
            - discounts: List of discount dictionaries with 'amount' and 'item_index'
            - shipping_cost: The cost of shipping
            - shipping_taxable: Boolean indicating if shipping is taxable
            - shipping_tax_rate: Tax rate for shipping
            - is_sales_before_tax: Boolean indicating if sales tax is applied before discounts
            - discount_is_taxable: Boolean indicating if discounts are taxable
    
    Returns:
        A dictionary containing the following keys:
            - total_tax: Total tax amount
            - total_amount: Total amount including tax
            - item_total: Total amount of items
            - discount_total: Total amount of discounts
            - shipping_tax: Tax amount on shipping
            - tax_breakdown: List of tax breakdown items
    """
    # Extract data from request
    items = data.get('items', [])
    discounts = data.get('discounts', [])
    shipping_cost = safe_decimal(data.get('shipping_cost'))
    shipping_taxable = data.get('shipping_taxable', False)
    shipping_tax_rate = safe_decimal(data.get('shipping_tax_rate'))
    is_sales_before_tax = data.get('is_sales_before_tax', False)
    discount_is_taxable = data.get('discount_is_taxable', True)

    total_tax = Decimal('0')
    total_amount = Decimal('0')
    item_total = Decimal('0')
    discount_total = Decimal('0')
    shipping_tax = Decimal('0')
    tax_breakdown = []

    # Validate shipping tax rate
    if shipping_cost > 0 and shipping_taxable and shipping_tax_rate == Decimal('0'):
        raise ValueError("Shipping is marked as taxable but no tax rate provided. Please enter a shipping tax rate.")
    
    # Initialize total tax
    total_tax = Decimal('0')

    # Calculate item totals
    for i, item in enumerate(items, 1):
        price = safe_decimal(item['price'])
        tax_rate = safe_decimal(item['tax_rate'])
        item_total += price

    # Calculate discount totals
    for discount in discounts:
        discount_total += safe_decimal(discount['amount'])

    # Subtract discounts from item_total for conditions 4 and 5
    if is_sales_before_tax and not discount_is_taxable:
        item_total -= discount_total

    # Process taxes based on conditions of is_sales_before_tax and discount_is_taxable
    if is_sales_before_tax:
        # Conditions 2 or 4 are both is_sales_before_tax is True
        if discount_is_taxable:
            # Condition 2 (When discount_is_taxable is True and is_sales_before_tax is True)
            for i, item in enumerate(items, 1):
                price = safe_decimal(item['price'])
                tax_rate = safe_decimal(item['tax_rate'])
                
                item_discounts = [d for d in discounts if d.get('item_index') == i]
                item_discount = sum(safe_decimal(d['amount']) for d in item_discounts)
                
                item_tax = calculate_tax(price - item_discount, tax_rate)
                discount_tax = calculate_tax(item_discount, tax_rate)
                
                total_tax += item_tax + discount_tax
                tax_breakdown.append({'item': f'Item {i}', 'tax': float(item_tax.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))})
                if item_discount > 0:
                    tax_breakdown.append({'item': f'Item {i} Discount', 'tax': float(discount_tax.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))})
        else:
            # Condition 4 (When discount_is_taxable is False and is_sales_before_tax is True)
            for i, item in enumerate(items, 1):
                price = safe_decimal(item['price'])
                tax_rate = safe_decimal(item['tax_rate'])
                
                item_discounts = [d for d in discounts if d.get('item_index') == i]
                item_discount = sum(safe_decimal(d['amount']) for d in item_discounts)
                
                item_tax = calculate_tax(price - item_discount, tax_rate)
                
                total_tax += item_tax
                tax_breakdown.append({'item': f'Item {i}', 'tax': float(item_tax.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))})
    else:
        # Conditions 1 or 3 are both is_sales_before_tax is False
        for i, item in enumerate(items, 1):
            price = safe_decimal(item['price'])
            tax_rate = safe_decimal(item['tax_rate'])
            
            item_tax = calculate_tax(price, tax_rate)
            total_tax += item_tax
            tax_breakdown.append({'item': f'Item {i}', 'tax': float(item_tax.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))})
            
            if discount_is_taxable:
                # Condition 1 (When discount_is_taxable is True and is_sales_before_tax is False)
                item_discounts = [d for d in discounts if d.get('item_index') == i]
                for discount in item_discounts:
                    discount_tax = calculate_tax(safe_decimal(discount['amount']), tax_rate)
                    total_tax += discount_tax
                    tax_breakdown.append({'item': f'Item {i} Discount', 'tax': float(discount_tax.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))})

    # Process shipping
    if shipping_cost > 0 and shipping_taxable:
        shipping_tax = calculate_tax(shipping_cost, shipping_tax_rate)
        total_tax += shipping_tax
        tax_breakdown.append({'item': 'Shipping', 'tax': float(shipping_tax.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP))})
    
    # Calculate final amount for case: is_sales_before_tax is True and discount_is_taxable is False
    if is_sales_before_tax and not discount_is_taxable:
        total_amount = item_total + total_tax + shipping_cost
    else:
        total_amount = item_total + total_tax + shipping_cost - discount_total
    
    if not discount_is_taxable:
        tax_breakdown = [tb for tb in tax_breakdown if 'Discount' not in tb['item']]

    # Round properly with Decimal module for accuracy
    result = {
        'item_total': float(item_total.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'discount_total': float(discount_total.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'shipping_cost': float(shipping_cost.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'shipping_tax': float(shipping_tax.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'total_tax': float(total_tax.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'total_amount': float(total_amount.quantize(Decimal('0.01'), rounding=ROUND_HALF_UP)),
        'tax_breakdown': tax_breakdown
    }

    return result

File: static/js/pages/tax_calculator.js
/**
 * tax_calculator.js - Tax Calculator functionality
 * Handles tax calculation logic, form interactions, and result display
 */

(function() {
    'use strict';
    
    // Module initialization
    document.addEventListener('DOMContentLoaded', function() {
        TaxCalculator.init();
    });
    
    const TaxCalculator = {
        // State variables
        itemCount: 0,
        discountCount: 0,
        
        init() {
            this.cacheElements();
            this.initializeCounters();
            this.bindEvents();
        },
        
        cacheElements() {
            this.elements = {
                form: document.getElementById('taxForm'),
                calculatorType: document.getElementById('calculator-type'),
                itemsContainer: document.getElementById('itemsContainer'),
                discountsContainer: document.getElementById('discountsContainer'),
                shippingTaxable: document.querySelector('select[name="shipping_taxable"]'),
                shippingTaxRateSection: document.getElementById('shipping_tax_rate_section'),
                addItemButton: document.getElementById('addItemBtn'),
                addDiscountButton: document.getElementById('addDiscountBtn'),
                calculateButton: document.querySelector('#taxForm button[type="submit"]')
            };
        },
        
        initializeCounters() {
            // Get initial counts from the form or default values
            const itemElements = document.querySelectorAll('.item');
            const discountElements = document.querySelectorAll('.discount');
            
            this.itemCount = itemElements.length || 1;
            this.discountCount = discountElements.length || 0;
        },
        
        bindEvents() {
            if (this.elements.form) {
                // Remove inline event handlers and replace with proper listeners
                this.elements.form.addEventListener('submit', this.handleFormSubmit.bind(this));
            }
            
            if (this.elements.calculatorType) {
                this.elements.calculatorType.removeAttribute('onchange');
                this.elements.calculatorType.addEventListener('change', this.handleCalculatorChange.bind(this));
            }
            
            if (this.elements.shippingTaxable) {
                this.elements.shippingTaxable.removeAttribute('onchange');
                this.elements.shippingTaxable.addEventListener('change', this.toggleShippingTaxRate.bind(this));
            }
            
            if (this.elements.addItemButton) {
                this.elements.addItemButton.removeAttribute('onclick');
                this.elements.addItemButton.addEventListener('click', this.addItem.bind(this));
            }
            
            if (this.elements.addDiscountButton) {
                this.elements.addDiscountButton.removeAttribute('onclick');
                this.elements.addDiscountButton.addEventListener('click', this.addDiscount.bind(this));
            }
            
            if (this.elements.calculateButton) {
                this.elements.calculateButton.addEventListener('click', this.handleFormSubmit.bind(this));
            }
        },
        
        /**
         * Handle calculator type change
         * @param {Event} event - Change event
         */
        handleCalculatorChange(event) {
            const selectElement = event.target;
            if (selectElement.value === 'canada') {
                window.location.href = selectElement.dataset.caCalculatorUrl;
            } else {
                // Stay on current page for US calculator
                return;
            }
        },
        
        /**
         * Add a new item row to the form
         */
        addItem() {
            this.itemCount++;
            const newItem = document.createElement('div');
            newItem.classList.add('item');
            newItem.innerHTML = `
                <label for="item_price_${this.itemCount}">Item ${this.itemCount} Price:</label>
                <input type="number" step="0.0001" name="item_price_${this.itemCount}" required>
                <label for="item_tax_rate_${this.itemCount}">Tax Rate (%):</label>
                <input type="number" step="0.0001" name="item_tax_rate_${this.itemCount}" required>
                <button type="button" class="delete-item-btn" data-item-id="${this.itemCount}">Delete</button>
            `;

            // Find the Add Item button and insert the new item before it
            const addItemBtn = this.elements.addItemButton;
            addItemBtn.parentNode.insertBefore(newItem, addItemBtn);

            // Bind delete button event
            const deleteBtn = newItem.querySelector('.delete-item-btn');
            deleteBtn.addEventListener('click', () => this.deleteItem(newItem));
        },
        
        /**
         * Add a new discount row to the form
         */
        addDiscount() {
            this.discountCount++;
            const newDiscount = document.createElement('div');
            newDiscount.classList.add('discount');
            newDiscount.innerHTML = `
                <label for="discount_amount_${this.discountCount}">Discount ${this.discountCount} Amount:</label>
                <input type="number" step="0.0001" name="discount_amount_${this.discountCount}">

                <label for="discount_item_${this.discountCount}">Apply to Item:</label>
                <select name="discount_item_${this.discountCount}">
                    ${Array.from({length: this.itemCount}, (_, i) =>
                        `<option value="${i+1}">Item ${i+1}</option>`
                    ).join('')}
                </select>
                <button type="button" class="delete-discount-btn" data-discount-id="${this.discountCount}">Delete</button>
            `;

            // Find the Add Discount button and insert the new discount before it
            const addDiscountBtn = this.elements.addDiscountButton;
            addDiscountBtn.parentNode.insertBefore(newDiscount, addDiscountBtn);

            // Bind delete button event
            const deleteBtn = newDiscount.querySelector('.delete-discount-btn');
            deleteBtn.addEventListener('click', () => this.deleteDiscount(newDiscount));
        },
        
        /**
         * Delete an item row (must keep at least 1 item)
         * @param {HTMLElement} itemElement - The item element to delete
         */
        deleteItem(itemElement) {
            const itemElements = document.querySelectorAll('.item');
            
            // Cannot delete if only 1 item remains
            if (itemElements.length <= 1) {
                alert('You must have at least one item.');
                return;
            }
            
            itemElement.remove();
        },
        
        /**
         * Delete a discount row (can delete all)
         * @param {HTMLElement} discountElement - The discount element to delete
         */
        deleteDiscount(discountElement) {
            discountElement.remove();
        },
        
        /**
         * Toggle shipping tax rate section visibility
         */
        toggleShippingTaxRate() {
            if (!this.elements.shippingTaxable || !this.elements.shippingTaxRateSection) return;
            
            const shippingTaxable = this.elements.shippingTaxable.value;
            this.elements.shippingTaxRateSection.style.display = shippingTaxable === 'Y' ? 'block' : 'none';
        },
        
        /**
         * Handle form submission
         * @param {Event} event - Submit event
         */
        handleFormSubmit(event) {
            event.preventDefault();
            this.calculateTaxes();
        },
        
        /**
         * Calculate taxes based on form inputs
         */
        calculateTaxes() {
            console.log("Calculating taxes...");
            
            // Get form data
            const items = [];
            let currentItemCount = 0;
            
            // Gather all item inputs
            while (document.querySelector(`input[name="item_price_${currentItemCount + 1}"]`)) {
                currentItemCount++;
                const priceInput = document.querySelector(`input[name="item_price_${currentItemCount}"]`);
                const taxRateInput = document.querySelector(`input[name="item_tax_rate_${currentItemCount}"]`);
                
                if (priceInput && taxRateInput) {
                    items.push({
                        price: parseFloat(priceInput.value) || 0,
                        tax_rate: parseFloat(taxRateInput.value) || 0
                    });
                }
            }
            
            // Gather all discount inputs
            const discounts = [];
            let currentDiscountCount = 0;
            
            while (document.querySelector(`input[name="discount_amount_${currentDiscountCount + 1}"]`)) {
                currentDiscountCount++;
                const discountInput = document.querySelector(`input[name="discount_amount_${currentDiscountCount}"]`);
                const itemIndexInput = document.querySelector(`select[name="discount_item_${currentDiscountCount}"]`);
                
                if (discountInput) {
                    discounts.push({
                        amount: parseFloat(discountInput.value) || 0,
                        item_index: itemIndexInput ? parseInt(itemIndexInput.value) : 1
                    });
                }
            }
            
            // Store the current form state in hidden fields to preserve it
            this.updateHiddenFields(currentItemCount, currentDiscountCount);
            
            // Get shipping inputs
            const shippingCost = parseFloat(document.querySelector('input[name="shipping_cost"]').value) || 0;
            const shippingTaxable = document.querySelector('select[name="shipping_taxable"]').value === 'Y';
            const shippingTaxRate = parseFloat(document.querySelector('input[name="shipping_tax_rate"]').value) || 0;
            
            // Get global settings
            const isSalesBeforeTax = document.querySelector('input[name="is_sales_before_tax"]').checked;
            const discountIsTaxable = document.querySelector('input[name="discount_is_taxable"]').checked;
            
            console.log("Items:", items);
            console.log("Discounts:", discounts);
            console.log("Settings:", { isSalesBeforeTax, discountIsTaxable, shippingCost, shippingTaxable, shippingTaxRate });
            
            // Calculate tax
            const result = this.calculateTax({
                items,
                discounts,
                shipping_cost: shippingCost,
                shipping_taxable: shippingTaxable,
                shipping_tax_rate: shippingTaxRate,
                is_sales_before_tax: isSalesBeforeTax,
                discount_is_taxable: discountIsTaxable
            });
            
            console.log("Result:", result);
            
            // Display results
            this.displayResults(result);
        },
        
        /**
         * Update hidden fields to preserve form state
         * @param {number} itemCount - Current item count
         * @param {number} discountCount - Current discount count
         */
        updateHiddenFields(itemCount, discountCount) {
            // Update item count hidden field
            let formStateField = document.querySelector('input[name="item_count"]');
            if (!formStateField) {
                formStateField = document.createElement('input');
                formStateField.type = 'hidden';
                formStateField.name = 'item_count';
                this.elements.form.appendChild(formStateField);
            }
            formStateField.value = itemCount;
        
            // Update discount count hidden field
            let discountCountField = document.querySelector('input[name="discount_count"]');
            if (!discountCountField) {
                discountCountField = document.createElement('input');
                discountCountField.type = 'hidden';
                discountCountField.name = 'discount_count';
                this.elements.form.appendChild(discountCountField);
            }
            discountCountField.value = discountCount;
        },
        
        /**
         * Calculate tax based on provided data
         * @param {Object} data - Tax calculation data
         * @returns {Object} - Tax calculation results
         */
        calculateTax(data) {
            const items = data.items || [];
            const discounts = data.discounts || [];
            const shippingCost = data.shipping_cost || 0;
            const shippingTaxable = data.shipping_taxable || false;
            const shippingTaxRate = data.shipping_tax_rate || 0;
            const isSalesBeforeTax = data.is_sales_before_tax || false;
            const discountIsTaxable = data.discount_is_taxable || false;
            
            let totalTax = 0;
            let totalAmount = 0;
            let itemTotal = 0;
            let discountTotal = 0;
            let shippingTax = 0;
            let taxBreakdown = [];
            
            // Calculate item totals
            for (const item of items) {
                itemTotal += item.price;
            }
            
            // Calculate discount totals
            for (const discount of discounts) {
                discountTotal += discount.amount;
            }
            
            // Adjust item_total for display in conditions 4 and 5
            let adjustedItemTotal = itemTotal;
            if (isSalesBeforeTax && !discountIsTaxable) {
                adjustedItemTotal = itemTotal - discountTotal;
            }
            
            // Process taxes based on conditions
            if (isSalesBeforeTax) {
                // Conditions 2 or 4
                if (discountIsTaxable) {
                    // Condition 2
                    for (let i = 0; i < items.length; i++) {
                        const item = items[i];
                        const itemDiscounts = discounts.filter(d => d.item_index === i + 1);
                        const itemDiscount = itemDiscounts.reduce((sum, d) => sum + d.amount, 0);
                        
                        const itemTax = (item.price - itemDiscount) * (item.tax_rate / 100);
                        const discountTax = itemDiscount * (item.tax_rate / 100);
                        
                        totalTax += itemTax + discountTax;
                        taxBreakdown.push({
                            item: `Item ${i + 1}`,
                            tax: Math.round(itemTax * 100) / 100
                        });
                        
                        if (itemDiscount > 0) {
                            taxBreakdown.push({
                                item: `Item ${i + 1} Discount`,
                                tax: Math.round(discountTax * 100) / 100
                            });
                        }
                    }
                } else {
                    // Adapted the Condition 4 from test_tools
                    for (let i = 0; i < items.length; i++) {
                        const item = items[i];
                        const itemDiscounts = discounts.filter(d => d.item_index === i + 1);
                        const itemDiscount = itemDiscounts.reduce((sum, d) => sum + d.amount, 0);
                        
                        const itemTax = (item.price - itemDiscount) * (item.tax_rate / 100);
                        
                        totalTax += itemTax;
                        taxBreakdown.push({
                            item: `Item ${i + 1}`,
                            tax: Math.round(itemTax * 100) / 100
                        });
                    }
                }
            } else {
                // Adapted the Conditions 1 or 3 from test_tools
                for (let i = 0; i < items.length; i++) {
                    const item = items[i];
                    const itemTax = item.price * (item.tax_rate / 100);
                    totalTax += itemTax;
                    taxBreakdown.push({
                        item: `Item ${i + 1}`,
                        tax: Math.round(itemTax * 100) / 100
                    });
                    
                    if (discountIsTaxable) {
                        // Adapted the Condition 1 from test_tools
                        const itemDiscounts = discounts.filter(d => d.item_index === i + 1);
                        for (const discount of itemDiscounts) {
                            const discountTax = discount.amount * (item.tax_rate / 100);
                            totalTax += discountTax;
                            taxBreakdown.push({
                                item: `Item ${i + 1} Discount`,
                                tax: Math.round(discountTax * 100) / 100
                            });
                        }
                    }
                }
            }
            
            // Process shipping
            if (shippingCost > 0 && shippingTaxable) {
                shippingTax = shippingCost * (shippingTaxRate / 100);
                totalTax += shippingTax;
                taxBreakdown.push({
                    item: 'Shipping',
                    tax: Math.round(shippingTax * 100) / 100
                });
            }
            
            // Calculate final amount using the correct formula based on conditions
            if (isSalesBeforeTax && !discountIsTaxable) {
                // For conditions 4 and 5
                totalAmount = adjustedItemTotal + totalTax + shippingCost;
            } else {
                // For other conditions
                totalAmount = itemTotal + totalTax + shippingCost - discountTotal;
            }
            
            // Filter out discount breakdowns for non-taxable discounts
            if (!discountIsTaxable) {
                taxBreakdown = taxBreakdown.filter(tb => !tb.item.includes('Discount'));
            }
            
            return {
                item_total: Math.round(adjustedItemTotal * 100) / 100,
                discount_total: Math.round(discountTotal * 100) / 100,
                shipping_cost: Math.round(shippingCost * 100) / 100,
                shipping_tax: Math.round(shippingTax * 100) / 100,
                total_tax: Math.round(totalTax * 100) / 100,
                total_amount: Math.round(totalAmount * 100) / 100,
                tax_breakdown: taxBreakdown
            };
        },
        
        /**
         * Display calculation results
         * @param {Object} result - Tax calculation results
         */
        displayResults(result) {
            // Find or create the result container
            let resultContainer = document.querySelector('.result-container');
            
            // If no container exists, create one
            if (!resultContainer) {
                console.log("Creating result container");
                resultContainer = document.createElement('div');
                resultContainer.className = 'result-container';
                document.querySelector('.tax-calculator').appendChild(resultContainer);
            }
            
            // Generate result HTML
            let html = `
                <h2>Results</h2>
                <p><strong>Item Total:</strong> $${result.item_total.toFixed(2)}</p>
                <p><strong>Discount Total:</strong> $${result.discount_total.toFixed(2)}</p>
                <p><strong>Shipping Cost:</strong> $${result.shipping_cost.toFixed(2)}</p>
                <p><strong>Total Tax:</strong> $${result.total_tax.toFixed(2)}</p>
                <p><strong>Total Amount:</strong> $${result.total_amount.toFixed(2)}</p>
                
                <div class="tax-breakdown">
                    <h3>Tax Breakdown:</h3>
                    <ul>
            `;
            
            for (const tax of result.tax_breakdown) {
                html += `<li>${tax.item}: $${tax.tax.toFixed(2)}</li>`;
            }
            
            html += `
                    </ul>
                </div>
            `;
            
            resultContainer.innerHTML = html;
        }
    };
})();
